File: debugoverlay/src/main/java/com/ms/square/debugoverlay/OverlayViewManager.java
package com.ms.square.debugoverlay;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.WindowManager.LayoutParams.FIRST_SYSTEM_WINDOW;
import static android.view.WindowManager.LayoutParams.TYPE_APPLICATION_ATTACHED_DIALOG;
import static android.view.WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.net.Uri;
import android.os.IBinder;
import android.provider.Settings;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;

import java.util.Collections;
import java.util.List;

class OverlayViewManager {

    private static final String TAG = "OverlayViewManager";

    private final Context context;
    private final DebugOverlay.Config config;
    private final WindowManager windowManager;

    private List<OverlayModule<?>> overlayModules = Collections.emptyList();
    private ViewGroup rootView;

    private boolean overlayPermissionRequested;

    public OverlayViewManager(@NonNull Context context, @NonNull DebugOverlay.Config config) {
        this.context = context;
        this.config = config;
        this.windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
    }

    public void setOverlayModules(@NonNull List<OverlayModule<?>> overlayModules) {
        this.overlayModules = overlayModules;
    }

    public void showDebugSystemOverlay() {
        if (config.isAllowSystemLayer() && rootView == null) {
            if (!canDrawOnSystemLayer(context, getWindowTypeForOverlay(true))) {
                Toast.makeText(context, R.string.debugoverlay_overlay_permission_prompt, Toast.LENGTH_LONG).show();
                requestDrawOnSystemLayerPermission(context);
                overlayPermissionRequested = true;
                return;
            }

            overlayPermissionRequested = false;

            rootView = createRoot();

            int layoutParamsWidth = WindowManager.LayoutParams.WRAP_CONTENT;

            for (OverlayModule<?> overlayModule : overlayModules) {
                View view = overlayModule.createView(rootView, config.getTextColor(), config.getTextSize(), config.getTextAlpha());
                if (view.getParent() == null) {
                    if (view.getLayoutParams() != null && view.getLayoutParams().width == MATCH_PARENT) {
                        layoutParamsWidth = WindowManager.LayoutParams.MATCH_PARENT;
                    }
                    rootView.addView(view);
                }
            }

            WindowManager.LayoutParams params = createLayoutParams(config.isAllowSystemLayer(), layoutParamsWidth, null);
            windowManager.addView(rootView, params);
        }
    }

    public void hideDebugSystemOverlay() {
        if (config.isAllowSystemLayer() && rootView != null) {
            windowManager.removeView(rootView);
            rootView = null;
        }
    }

    public boolean isSystemOverlayShown() {
        return rootView != null;
    }

    public boolean isOverlayPermissionRequested() {
        return overlayPermissionRequested;
    }

    public OverlayViewAttachStateChangeListener createAttachStateChangeListener() {
        return new OverlayViewAttachStateChangeListener();
    }

    // for layoutParams.gravity assignment
    @SuppressLint("WrongConstant")
    private WindowManager.LayoutParams createLayoutParams(boolean allowSystemLayer, int width, IBinder windowToken) {
        WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams();
        layoutParams.width = width;
        layoutParams.height = WindowManager.LayoutParams.WRAP_CONTENT;
        if (windowToken != null) {
            layoutParams.token = windowToken;
        }
        layoutParams.type = getWindowTypeForOverlay(allowSystemLayer);
        layoutParams.flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE | WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE;
        layoutParams.format = PixelFormat.TRANSLUCENT;
        layoutParams.gravity = config.getPosition().getGravity();
        return layoutParams;
    }

    private ViewGroup createRoot() {
        LinearLayout overlayRoot = new LinearLayout(context);
        overlayRoot.setOrientation(LinearLayout.VERTICAL);
        if (config.getBgColor() != Color.TRANSPARENT) {
            overlayRoot.setBackgroundColor(config.getBgColor());
        }
        return overlayRoot;
    }

    class OverlayViewAttachStateChangeListener implements View.OnAttachStateChangeListener {

        private ViewGroup _rootView;

        public void onActivityResumed() {
            if (_rootView != null && _rootView.getChildCount() > 0) {
                if (DebugOverlay.DEBUG) {
                    Log.i(TAG, "overlay views recreated on Activity's onResume");
                }
                _rootView.removeAllViews();
                for (OverlayModule<?> overlayModule : overlayModules) {
                    View view = overlayModule.createView(_rootView, config.getTextColor(), config.getTextSize(), config.getTextAlpha());
                    if (view.getParent() == null) {
                        _rootView.addView(view);
                    }
                }
                // force-update recreated views with the latest data
                for (OverlayModule<?> overlayModule : overlayModules) {
                    overlayModule.notifyObservers();
                }
            }
        }

        @Override
        public void onViewAttachedToWindow(@NonNull View v) {
            if (DebugOverlay.DEBUG) {
                Log.i(TAG, "onViewAttachedToWindow");
            }
            _rootView = createRoot();
            int layoutParamsWidth = WindowManager.LayoutParams.WRAP_CONTENT;
            for (OverlayModule<?> overlayModule : overlayModules) {
                View view = overlayModule.createView(_rootView, config.getTextColor(), config.getTextSize(), config.getTextAlpha());
                if (view.getParent() == null) {
                    if (view.getLayoutParams() != null && view.getLayoutParams().width == MATCH_PARENT) {
                        layoutParamsWidth = WindowManager.LayoutParams.MATCH_PARENT;
                    }
                    _rootView.addView(view);
                }
            }

            windowManager.addView(_rootView, createLayoutParams(config.isAllowSystemLayer(),
                    layoutParamsWidth, v.getWindowToken()));
        }

        @Override
        public void onViewDetachedFromWindow(@NonNull View v) {
            if (DebugOverlay.DEBUG) {
                Log.i(TAG, "onViewDetachedFromWindow");
            }
            windowManager.removeViewImmediate(_rootView);
            v.removeOnAttachStateChangeListener(this);
        }
    }

    public static void requestDrawOnSystemLayerPermission(@NonNull Context context) {
        Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                Uri.parse("package:" + context.getPackageName()));
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        context.startActivity(intent);
    }

    public static boolean canDrawOnSystemLayer(@NonNull Context context, int systemWindowType) {
        if (isSystemLayer(systemWindowType)) {
            return Settings.canDrawOverlays(context);
        }
        return true;
    }

    public static boolean isSystemLayer(int windowType) {
        return windowType >= FIRST_SYSTEM_WINDOW;
    }

    public static int getWindowTypeForOverlay(boolean allowSystemLayer) {
        if (allowSystemLayer) {
            return TYPE_APPLICATION_OVERLAY;
        } else {
            // make layout of the window happens as that of a top-level window, not as a child of its container
            return TYPE_APPLICATION_ATTACHED_DIALOG;
        }
    }
}

File: debugoverlay-core/src/main/kotlin/com/ms/square/debugoverlay/OverlayViewManager.kt
package com.ms.square.debugoverlay

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.PixelFormat
import android.os.IBinder
import android.provider.Settings
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Toast
import androidx.compose.ui.platform.ComposeView
import androidx.core.net.toUri
import androidx.lifecycle.setViewTreeLifecycleOwner
import androidx.savedstate.setViewTreeSavedStateRegistryOwner
import com.ms.square.debugoverlay.internal.ui.DebugOverlayPanel

internal class OverlayViewManager(
  private val context: Context,
  private val config: DebugOverlay.Config,
) {
    private val windowManager: WindowManager =
        context.getSystemService(Context.WINDOW_SERVICE) as WindowManager

    private var rootView: ViewGroup? = null
    private var overlayPermissionRequested = false

    fun showDebugSystemOverlay() {
        if (config.isAllowSystemLayer && rootView == null) {
            if (!canDrawOnSystemLayer(context, getWindowTypeForOverlay(true))) {
                Toast.makeText(
                    context,
                    R.string.debugoverlay_overlay_permission_prompt,
                    Toast.LENGTH_LONG
                ).show()
                requestDrawOnSystemLayerPermission(context)
                overlayPermissionRequested = true
                return
            }

            overlayPermissionRequested = false

            rootView = createRoot()

            val params = createLayoutParams(config.isAllowSystemLayer, null)
            windowManager.addView(rootView, params)
        }
    }

    fun hideDebugSystemOverlay() {
        if (config.isAllowSystemLayer && rootView != null) {
            windowManager.removeView(rootView)
            rootView = null
        }
    }

    fun isSystemOverlayShown(): Boolean = rootView != null

    fun isOverlayPermissionRequested(): Boolean = overlayPermissionRequested

    fun createAttachStateChangeListener(): OverlayViewAttachStateChangeListener {
        return OverlayViewAttachStateChangeListener()
    }

    @SuppressLint("WrongConstant")
    private fun createLayoutParams(
      allowSystemLayer: Boolean,
      windowToken: IBinder?,
    ): WindowManager.LayoutParams {
        return WindowManager.LayoutParams().apply {
            width = WindowManager.LayoutParams.WRAP_CONTENT
            height = WindowManager.LayoutParams.WRAP_CONTENT
            if (windowToken != null) {
                token = windowToken
            }
            type = getWindowTypeForOverlay(allowSystemLayer)
            flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or
                    WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
            format = PixelFormat.TRANSLUCENT
            gravity = Gravity.BOTTOM or Gravity.START
        }
    }

    private fun createRoot(): ViewGroup {
        return ComposeView(context).apply {
            // Create and attach a synthetic lifecycle for the overlay
            val lifecycleOwner = OverlayLifecycleOwner()
            setViewTreeLifecycleOwner(lifecycleOwner)
            setViewTreeSavedStateRegistryOwner(lifecycleOwner)

            // Start the lifecycle
            lifecycleOwner.onCreate()

            setContent {
                DebugOverlayPanel(
                  onClick = {
                    //TODO: Navigate to detailed performance screen
                  }
                )
            }

            // Move lifecycle to STARTED/RESUMED when the view is attached
            addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
                override fun onViewAttachedToWindow(v: View) {
                    lifecycleOwner.onStart()
                    lifecycleOwner.onResume()
                }

                override fun onViewDetachedFromWindow(v: View) {
                    lifecycleOwner.onPause()
                    lifecycleOwner.onStop()
                    lifecycleOwner.onDestroy()
                }
            })
        }
    }

    inner class OverlayViewAttachStateChangeListener : View.OnAttachStateChangeListener {
        private var _rootView: ViewGroup? = null

        fun onActivityResumed() {
//            _rootView?.let { root ->
//              if (overlayModules.isNotEmpty()) {
//                // force-update recreated views with the latest data
//                for (overlayModule in overlayModules) {
//                  overlayModule.notifyObservers()
//                }
//              }
//            }
        }

        override fun onViewAttachedToWindow(v: View) {
            if (DebugOverlay.DEBUG) {
                Log.i(TAG, "onViewAttachedToWindow")
            }
            _rootView = createRoot()
            windowManager.addView(
                _rootView,
                createLayoutParams(config.isAllowSystemLayer, v.windowToken)
            )
        }

        override fun onViewDetachedFromWindow(v: View) {
            if (DebugOverlay.DEBUG) {
                Log.i(TAG, "onViewDetachedFromWindow")
            }
            _rootView?.let { windowManager.removeViewImmediate(it) }
            v.removeOnAttachStateChangeListener(this)
        }
    }

    companion object {
        private const val TAG = "OverlayViewManager"

        @JvmStatic
        fun requestDrawOnSystemLayerPermission(context: Context) {
            val intent = Intent(
                Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
              "package:${context.packageName}".toUri()
            ).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            }
            context.startActivity(intent)
        }

        @JvmStatic
        fun canDrawOnSystemLayer(context: Context, systemWindowType: Int): Boolean {
            return if (isSystemLayer(systemWindowType)) {
                Settings.canDrawOverlays(context)
            } else {
                true
            }
        }

        @JvmStatic
        fun isSystemLayer(windowType: Int): Boolean {
            return windowType >= WindowManager.LayoutParams.FIRST_SYSTEM_WINDOW
        }

        @JvmStatic
        fun getWindowTypeForOverlay(allowSystemLayer: Boolean): Int {
            return if (allowSystemLayer) {
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            } else {
                // make layout of the window happens as that of a top-level window, not as a child of its container
                WindowManager.LayoutParams.TYPE_APPLICATION_ATTACHED_DIALOG
            }
        }
    }
}
